File: packages/sandboxd/src/protocol/keepalive.rs
//! Serialized keepalive protocol shared with the control/data plane.
//!
//! The daemon emits this shape to describe whether sandbox activity should keep
//! the instance alive and when the next heartbeat is expected.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct KeepaliveState {
    #[serde(rename = "type")]
    pub message_type: KeepaliveMessageType,
    pub ttl_ms: u64,
    pub active: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum KeepaliveMessageType {
    #[serde(rename = "keepalive.state")]
    State,
}

File: packages/sandboxd/protocol/keepalive.go
package protocol

import "fmt"

type KeepaliveMessageType string

const KeepaliveMessageState KeepaliveMessageType = "keepalive.state"

type KeepaliveState struct {
	MessageType KeepaliveMessageType `json:"type"`
	TTLMS       uint64               `json:"ttlMs"`
	Active      bool                 `json:"active"`
}

func DecodeKeepaliveState(data []byte) (KeepaliveState, error) {
	var state KeepaliveState
	if err := decodeStrict(data, &state); err != nil {
		return KeepaliveState{}, err
	}
	if state.MessageType != KeepaliveMessageState {
		return KeepaliveState{}, fmt.Errorf("unsupported keepalive message type: %s", state.MessageType)
	}
	return state, nil
}
